aio/local_packages_util.bzl

```
@@ -1,47 +1,119 @@
 load("@build_bazel_rules_nodejs//internal/linker:npm_link.bzl", "npm_link")
+load("@build_bazel_rules_nodejs//:providers.bzl", "LinkablePackageInfo")
+load("@build_bazel_rules_nodejs//internal/linker:link_node_modules.bzl", "LinkerPackageMappingInfo")
 
-def link_local_packages(deps):
-    """Stamp npm_link targets for packages in deps that has a local package equivalent.
+def _is_angular_dep(dep):
+    """Check if a dep , e.g., @aio_npm//@angular/core corresonds to a local Angular pacakge."""
+    if dep.startswith("@aio_npm//"):
+        label = Label(dep)
+        if label.package in ALL_PACKAGES:
+            return True
+    return False
+
+def _angular_dep_to_pkg_name(dep):
+    """E.g., @aio_npm//@angular/core => '@angular/core'"""
+    label = Label(dep)
+    return label.package
+
+def link_local_packages(all_aio_deps):
+    """Stamp targets needed for building AIO against local angular packages.
+
+    Stamps targets that link Angular packages, as well as targets to be used
+    in place of any deps required to build and test AIO. These targets filter
+    out any transitive deps on the npm packages and must be used in place of
+    any original list of deps.
+
+    Use the helper `substitute_local_package_deps()` to translate a list of deps
+    to the equivalent "filtered" target.
 
     Args:
-        deps: list of npm dependency labels
+        all_aio_deps: label list of all deps required to build and test AIO
     """
-    for dep in deps:
-        if dep.startswith("@aio_npm//"):
-            label = Label(dep)
-            if label.package in ALL_PACKAGES:
-                npm_link(
-                    name = _npm_link_name(dep),
-                    target = to_package_label(label.package),
-                    package_name = label.package,
-                    package_path = native.package_name(),
-                    tags = ["manual"],
-                )
-
-def substitute_local_packages(deps):
-    """Substitute npm dependencies for their local npm_link equivalent.
-
-    Assumes that link_local_packages() was already called on these dependencies.
-    Dependencies that are not associated with a local package are left alone.
+
+    aio_angular_deps = [dep for dep in all_aio_deps if _is_angular_dep(dep)]
+    angular_packages = [_angular_dep_to_pkg_name(dep) for dep in aio_angular_deps]
+
+    # Link local angular packages in place of their npm equivalent
+    for dep in aio_angular_deps:
+        pkg_name = _angular_dep_to_pkg_name(dep)
+        npm_link(
+            name = _npm_link_name(pkg_name),
+            target = to_package_label(pkg_name),
+            package_name = pkg_name,
+            package_path = native.package_name(),
+            tags = ["manual"],
+        )
+
+    # Special case deps that must be testonly
+    testonly_deps = [
+        "@aio_npm//@angular/build-tooling/bazel/browsers/chromium",
+    ]
+
+    # Stamp a corresponding target for each AIO dep that filters out transitive
```

```suggestion
    # Create a corresponding target for each AIO dep that filters out transitive
```

mostly to avoid ambiguity with stamping in the context of bazel